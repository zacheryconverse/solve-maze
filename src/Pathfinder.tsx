import React, { Component } from "react";
import Maze from './Maze'

type MyState = {
  maze: any;
  startCol: number;
  startRow: number;
  directions: string[];
}

class Pathfinder extends Component {
  state: MyState = {
    maze: [
      [0, 0, 0, 0, 0],
      [1, 1, 1, 0, 0],
      [0, 0, 1, 0, 0],
      [0, 1, 1, 0, 0],
      [0, 1, 0, 1, "x"],
      [0, 1, 1, 1, 0],
    ],
    startCol: 1,
    startRow: 0,
    directions: []
  };

  findDirection = (prev: number[], curr: number[]) => {
    const pCol = prev[0];
    const pRow = prev[1];
    const cCol = curr[0];
    const cRow = curr[1];
    if (pCol === cCol && pRow < cRow) return 'ðŸ‘‡ Down';
    if (pCol === cCol && pRow > cRow) return 'ðŸ‘† Up';
    if (pCol > cCol && pRow === cRow) return 'ðŸ‘ˆ Left';
    if (pCol < cCol && pRow === cRow) return 'ðŸ‘‰ Right';
  };

  movesCount!: number;
  maze!: number[] | string[];
  coordinates: number[] = [];
  perform!: ((col: number, row: number) => void);

  solveMaze = (maze: number[] | string[]) => {
    const { startCol, startRow, directions } = this.state;
    this.movesCount = 0;
    this.maze = maze;
    this.coordinates = [startCol, startRow];

    this.perform = (col: number, row: number) => {
      if (this.maze[col][row] === 1) {
        console.log(
          this.movesCount ?
          this.findDirection(this.coordinates, [col, row]) :
          'ðŸŽ‰ Start'
        );
        this.movesCount++;
        this.coordinates = [col, row];
        this.maze[col][row] = "/";
        this.setState({
          directions: [...directions, `${col}-${row}`]
        });

        if (col < this.maze.length - 1) this.perform(col + 1, row);
        if (col > 0) this.perform(col - 1, row);
        if (row < this.maze.length - 1) this.perform(col, row + 1);
        if (row > 0) this.perform(col, row - 1);
      } else if (this.maze[col][row] === "x") {
        console.log(
          `${this.findDirection(this.coordinates, [col, row])} ðŸŽ‰\n
          Finished in ${this.movesCount} moves!\n
          ---------------------
        `);
      }
    };
    this.perform(startCol, startRow);
  };

  render() {
    const { maze, startCol, startRow, directions } = this.state;
    return (
      <div className='page'>
        Solve the Maze!
        <Maze
          maze={maze}
          startCol={startCol}
          startRow={startRow}
          solveMaze={this.solveMaze}
          directions={directions}
        />
        <button
          onClick={() => this.solveMaze(maze)}
          className='btn'
        >
            Go
        </button>
      </div>
    );
  };
};

export default Pathfinder;
