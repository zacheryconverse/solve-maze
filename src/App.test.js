import React from 'react';
import ReactDOM from 'react-dom';
import { render, screen } from '@testing-library/react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import Pathfinder from './Pathfinder';

configure({ adapter: new Adapter() });

const app = shallow(<Pathfinder />, { disableLifecycleMethods: true });

it('renders without a hitch', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Pathfinder />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('<Pathfinder />', () => {
  beforeAll(() => {
    global.fetch = jest.fn();
  })

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Pathfinder />, { disableLifecycleMethods: true });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has a page', () => {
    expect(wrapper.find('.page')).toBeDefined();
  });

  it('has a maze', () => {
    expect(wrapper.find('.maze')).toBeDefined();
  });

  it('has a Maze component', () => {
    expect(wrapper.find('Maze')).toBeDefined();
  });

  test('renders solve greeting', () => {
    render(<Pathfinder />);
    const linkElement = screen.getByText(/Solve the Maze!/i);
    expect(linkElement).toBeInTheDocument();
  });

  test('the maze array has `x` in it', () => {
    expect(app.state().maze[4]).toContain('x');
    expect(new Set(app.state().maze[4])).toContain('x');
  });

  test('Should have `Go` button', () => {
    expect(wrapper.find('.btn').text()).toBe('Go');
  })
});

it('should be able to activate button with Space-bar', () => {
  const component = mount(<Pathfinder />);
  component
    .find('.btn')
    .simulate('keydown', { keyCode: 32 });
  expect(component).toMatchSnapshot();
  component.unmount();
});

const clickFn = jest.fn();
describe('Pathfinder', () => {
  it('button click should call solve function', () => {
    const component = shallow(<Pathfinder onClick={clickFn()} />);
    component
      .find('.btn')
      .simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});
