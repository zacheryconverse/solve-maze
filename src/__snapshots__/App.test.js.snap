// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pathfinder /> matches the snapshot 1`] = `
<div
  className="page"
>
  Solve the Maze! Open Console
  <Maze
    directions={Array []}
    maze={
      Array [
        Array [
          0,
          0,
          0,
          0,
          0,
        ],
        Array [
          1,
          1,
          1,
          0,
          0,
        ],
        Array [
          0,
          0,
          1,
          0,
          0,
        ],
        Array [
          0,
          1,
          1,
          0,
          0,
        ],
        Array [
          0,
          1,
          0,
          1,
          "x",
        ],
        Array [
          0,
          1,
          1,
          1,
          0,
        ],
      ]
    }
    solveMaze={[Function]}
    startCol={1}
    startRow={0}
  />
  <button
    className="btn"
    onClick={[Function]}
  >
    Go
  </button>
</div>
`;

exports[`should be able to activate button with Space-bar 1`] = `
<Pathfinder>
  <div
    className="page"
  >
    Solve the Maze! Open Console
    <Maze
      directions={Array []}
      maze={
        Array [
          Array [
            0,
            0,
            0,
            0,
            0,
          ],
          Array [
            1,
            1,
            1,
            0,
            0,
          ],
          Array [
            0,
            0,
            1,
            0,
            0,
          ],
          Array [
            0,
            1,
            1,
            0,
            0,
          ],
          Array [
            0,
            1,
            0,
            1,
            "x",
          ],
          Array [
            0,
            1,
            1,
            1,
            0,
          ],
        ]
      }
      solveMaze={[Function]}
      startCol={1}
      startRow={0}
    >
      <div
        className="maze"
        style={
          Object {
            "gridTemplateColumns": "repeat(6, 0fr)",
          }
        }
      >
        <div
          key="0"
        >
          <div
            className="cell
                zero"
            key="00"
          />
          <div
            className="cell
                zero"
            key="01"
          />
          <div
            className="cell
                zero"
            key="02"
          />
          <div
            className="cell
                zero"
            key="03"
          />
          <div
            className="cell
                zero"
            key="04"
          />
        </div>
        <div
          key="1"
        >
          <div
            className="cell
                start"
            key="10"
          />
          <div
            className="cell
                one"
            key="11"
          />
          <div
            className="cell
                one"
            key="12"
          />
          <div
            className="cell
                zero"
            key="13"
          />
          <div
            className="cell
                zero"
            key="14"
          />
        </div>
        <div
          key="2"
        >
          <div
            className="cell
                zero"
            key="20"
          />
          <div
            className="cell
                zero"
            key="21"
          />
          <div
            className="cell
                one"
            key="22"
          />
          <div
            className="cell
                zero"
            key="23"
          />
          <div
            className="cell
                zero"
            key="24"
          />
        </div>
        <div
          key="3"
        >
          <div
            className="cell
                zero"
            key="30"
          />
          <div
            className="cell
                one"
            key="31"
          />
          <div
            className="cell
                one"
            key="32"
          />
          <div
            className="cell
                zero"
            key="33"
          />
          <div
            className="cell
                zero"
            key="34"
          />
        </div>
        <div
          key="4"
        >
          <div
            className="cell
                zero"
            key="40"
          />
          <div
            className="cell
                one"
            key="41"
          />
          <div
            className="cell
                zero"
            key="42"
          />
          <div
            className="cell
                one"
            key="43"
          />
          <div
            className="cell
                finish"
            key="44"
          />
        </div>
        <div
          key="5"
        >
          <div
            className="cell
                zero"
            key="50"
          />
          <div
            className="cell
                one"
            key="51"
          />
          <div
            className="cell
                one"
            key="52"
          />
          <div
            className="cell
                one"
            key="53"
          />
          <div
            className="cell
                zero"
            key="54"
          />
        </div>
      </div>
    </Maze>
    <button
      className="btn"
      onClick={[Function]}
    >
      Go
    </button>
  </div>
</Pathfinder>
`;
